# file locations ----------------------------------------------------------
#global pval cutoff file names
pval_cutoff=0.1 #Onuchic use 0.1
gff_in_file='../downstream/input/human_analysis/gff_in.rds'
variant_HetCpG_file='../downstream/input/human_analysis/variant_HetCpG_new.rds'
GR_file='../downstream/output/human_analysis/CPEL_outputs/GRs_final1.rds'
GR_allele_file='../downstream/output/human_analysis/CPEL_outputs/GRs_allele_final1.rds'
hetCpG_gff_file='../downstream/input/hetCpG_gff_final1.rds'
GR_merge_file="../downstream/output/human_analysis/CPEL_outputs/GR_merge_final12_ls.rds"
CpG_hg19_file='../downstream/input/human_analysis/CpG_hg19.rds'
variant_HetCpG_meta_file='../downstream/output/human_analysis/CPEL_outputs/variant_HetCpG_meta_final1_ls.rds'
genomic_features_file="../downstream/input/human_analysis/genomic_features2020.rds"
NME_agnostic_file="../downstream/output/human_analysis/CPEL_outputs/NME_allele_agnostic_merge_20k_homogeneous_excluding_dMML2.rds"
MML_agnostic_file="../downstream/output/human_analysis/CPEL_outputs/MML_allele_agnostic_merge_20k_homogeneous2.rds"
motif_gene_file='../downstream/output/human_analysis/motif_analysis/motif_all_JASPAR_default.rds' #For all SNP
NME_agnostic_DNase_file="../downstream/output/human_analysis/CPEL_outputs/allele_agnostic_hg19_DNase_NME_homogeneous_excluding_dMML.rds"
MML_agnostic_DNase_file="../downstream/output/human_analysis/CPEL_outputs/allele_agnostic_hg19_DNase_MML_homogeneous.rds"
NME_agnostic_ASM_file="../downstream/output/human_analysis/CPEL_outputs/NME_agnostic_ASM.rds"
MML_agnostic_ASM_file="../downstream/output/human_analysis/CPEL_outputs/MML_agnostic_ASM.rds"
NME_agnostic_comp_file="../downstream/output/human_analysis/CPEL_outputs/NME_agnostic_comp.rds"
MML_agnostic_comp_file="../downstream/output/human_analysis/CPEL_outputs/MML_agnostic_comp.rds"
NME_agnostic_all_file="../downstream/output/human_analysis/CPEL_outputs/NME_agnostic_all.rds"
MML_agnostic_all_file="../downstream/output/human_analysis/CPEL_outputs/MML_agnostic_all.rds"
variant_HetCpG_meta_dt_file='../downstream/output/human_analysis/CpG_density/variant_HetCpG_meta_dt.rds'
variant_HetCpG_meta_dt_uq_file='../downstream/output/human_analysis/CpG_density/variant_HetCpG_meta_dt_relative_dNME2uq.rds'
figure_path='../downstream/output/graphs_tables/'
#This is from Ken
DNase_hg19_file='../downstream/input/human_analysis/DNase_hg19_250bp.rds'
control_hg19_file='../downstream/input/human_analysis/DNase_hg19_250bp_control.rds'
JASPAR_motif_hg19_file='../downstream/output/human_analysis/motif_analysis/motif_JASPAR_hg19.rds'
#This is from Ken
DNase_mm10_file='../downstream/input/mouse_analysis/DNase_mm10_peak_merge_250bp.rds'
control_mm10_file='../downstream/input/mouse_analysis/DNase_mm10_peak_merge_250bp_control.rds'
JASPAR_motif_mm10_file='../downstream/input/mouse_analysis/motif_analysis/motif_JASPAR_mm10.rds'
#Mouse analysis
MML_matrix_file='../downstream/output/mouse_analysis/CPEL_outputs/MML_matrix_mouse_all_dedup_N2_all_regions.rds'
NME_matrix_file='../downstream/output/mouse_analysis/CPEL_outputs/NME_matrix_mouse_all_dedup_N2_all_regions.rds'
CG_density_mouse='../downstream/output/mouse_analysis/CpG_density_NME/region_CG_exp.rds'
mouse_DNase_control_gff_file='../downstream/output/mouse_analysis/CPEL_inputs/mm10_allele_agnostic_analysis_DNase_control.gff'
mouse_compliment_gff_file='../downstream/output/mouse_analysis/CPEL_inputs/mm10_allele_agnostic_analysis_compliment.gff'
UC_in_matrix_ls_file='../downstream/output/mouse_analysis/CPEL_outputs/UC_matrix_ls_N2_all_regions.rds'
UC_in_matrix_cluster_file='../downstream/output/mouse_analysis/CPEL_outputs/UC_matrix_ls_N2_all_regions_clu.rds'
UC_in_MDS_all_file='../downstream/output/mouse_analysis/CPEL_outputs/UC_MDS_N2_all_regions.rds'
UC_in_MDS_all_P0_file='../downstream/output/mouse_analysis/CPEL_outputs/UC_MDS_N2_all_regions_P0.rds'
UC_in_MDS_all_P0_all_file='../downstream/output/mouse_analysis/CPEL_outputs/UC_MDS_N2_all_regions_all_P0.rds'

dir_cluster_in='../downstream/output/mouse_analysis/clustering/tissue_specific/'
dir_cluster_in_non_ts='../downstream/output/mouse_analysis/clustering/non_tissue_specific/'
dir_cluster_in_01=paste0(dir_cluster_in,'uc_01/')
dir_cluster_in_01_non_ts=paste0(dir_cluster_in_non_ts,'uc_01/')
UC_merge_file='../downstream/output/mouse_analysis/CPEL_outputs/UC_merge_all_regions.rds'
UC_merge_max_loc_file='../downstream/output/mouse_analysis/CPEL_outputs/UC_merge_all_regions_max_loc.rds'
UC_merge_max_loc_01_file='../downstream/output/mouse_analysis/CPEL_outputs/UC_merge_all_regions_max_loc_01.rds'
dir_out_cluster01=paste0(dir_cluster_in,'UC_0_1/cluster_assigned/')
dir_out_cluster01_non_ts=paste0(dir_cluster_in_non_ts,'UC_0_1/cluster_assigned/')
cluster_01_region_out_fn=paste0(dir_out_cluster01,'cluster_all_region_assignment_filtered_0_1.rds')
cluster_01_region_out_non_ts_fn=paste0(dir_out_cluster01,'cluster_all_region_assignment_filtered_non_ts_0_1.rds')
dmml_cor_file='../downstream/input/mouse_analysis/correlation_analysis/all_regions/fulldmmlcor_YQ.rds'
dnme_cor_file='../downstream/input/mouse_analysis/correlation_analysis/all_regions/fulldnmecor_YQ.rds'
dir_out_rds_correlation='../downstream/output/mouse_analysis/correlation/UC_01/'
bin_enhancer_rds='../downstream/input/mouse_analysis/enhancer_selection/bin_enhancer.rds'
bin_enhancer_bed='../downstream/input/mouse_analysis/enhancer_selection/bin_enhancer.bed'
GO_01_dir='../downstream/output/mouse_analysis/GO_analysis/kmeans_N17_10run_01/'
tissue_out_filtered_fn=paste0(dir_out_rds_correlation,'tissue_out_N17_kmeans_10run_filtered_all_region.rds')
enhancer_region_fn='../downstream/output/mouse_analysis/GO_analysis/all_regions_enchancer.rds'
GO_01_enhancer_fn=paste0(GO_01_dir,'GO_out_all_dMML_dNME_0rm_FC_N17_kmeans_10run_filtered_all_regions_01_enhancer.rds')
GO_01_promoter_fn=paste0(GO_01_dir,'GO_out_all_dMML_dNME_0rm_FC_N17_kmeans_10run_filtered_all_regions_01_promoter.rds')
gene_example_dir=paste0(GO_01_dir,'gene_examples/')
tissue_region_motif_all_regions_fn="../downstream/output/mouse_analysis/motif_analysis/tissue_region_motif_all_regions.rds"
#This is from Ken via slack
mouse_motif_dir='../downstream/output/mouse_analysis/motif_analysis/'
Ken_motif_folder='../downstream/input/mouse_analysis/motif_analysis/mouse_motif_enrichment_0526/'
region_motif_dir=paste0(mouse_motif_dir,'region_motif/')
cancer_input_dir='../downstream/input/human_analysis/cancer_SNP/'
cosmic_pan_mutation_fn='../downstream/input/human_analysis/pan_cancer_mutation_coding_non_coding_COSMIC.rds'
cosmic_pan_mutation_fn_hg19='../downstream/input/human_analysis/pan_cancer_mutation_coding_non_coding_COSMIC_hg19.rds'
motif_locus_bed_dir='../downstream/output/mouse_analysis/motif_analysis/motif_locus_bed/'
enhancer_motif_all_dNME_fn=paste0(mouse_motif_dir,'enhancer_regions_motif_dNME_all2.rds')
enhancer_motif_all_dMML_fn=paste0(mouse_motif_dir,'enhancer_regions_motif_dMML_all2.rds')
GWAS_prep_dir='../downstream/output/mouse_analysis/GWAS_prep/'
chip_atlas_dir='../downstream/input/mouse_analysis/motif_analysis/chipatlas/'
ChiP_motif_dir=paste0(mouse_motif_dir,'motif_Chip_rds/')
scRNA_dir="../downstream/input/human_analysis/NME_expression_var/scRNA/"
NME_MAV_human_out_dir='../downstream/output/human_analysis/NME_MAV/'
ICGC_mutation_fn='../downstream/input/human_analysis/human_variant_passenger_sm.rds'
NME_in_CG_fn='../downstream/input/mouse_analysis/human_conservation/NME_matrix_N2_CG.rds'
NME_mouse_MAV_fn='../downstream/output/mouse_analysis/NME_in_limb_ENOCD3_imputed.rds'
Ken_motif_out_dir='../downstream/output/mouse_analysis/mouse_motif_Ken/'
dir_scRNA_mouse='../downstream/input/mouse_analysis/Mouse_C1/'
NME_mouse_MAV_enhancer_fn='../downstream/output/mouse_analysis/NME_in_limb_ENOCD3_imputed_enhancer.rds'
compliment_MDS_dir_P0='../downstream/data/UC_MDS_day0/'
DNase_conrol_MDS_dir='../downstream/data/DNase_control_PRC_MDS_mouse/'
tissue_specific_out='../downstream/output/mouse_analysis/tissue_specific_enhancer/'
H3K27ac_output_dt_fn=paste0(tissue_specific_out,'H3K27ac_output_dt.rds')
hg19_control_NME_fn="../downstream/output/human_analysis/Ken_motif/homogeneous/human_motif_control_median_NME.rds"
hg19_DNase_NME_fn="../downstream/output/human_analysis/Ken_motif/homogeneous/human_motif_DNase_median_NME.rds"
hg19_control_MML_fn="../downstream/output/human_analysis/Ken_motif/homogeneous/human_motif_control_median_MML.rds"
hg19_DNase_MML_fn="../downstream/output/human_analysis/Ken_motif/homogeneous/human_motif_DNase_median_MML.rds"
JSD_in_fn='../downstream/output/mouse_analysis/model_QC/JSD_agnostic_mouse_matrix_dedup_N2_all_merged_ls.rds'
UC_in_QC_fn='../downstream/output/mouse_analysis/model_QC/UC_JSD_QC.rds'